File: frontend/flutter_application_1/lib/models/meal_plan.dart
class MealPlan {
  final String id;
  final String userId;
  final DateTime planDate;
  final int durationDays;
  final List<Meal> meals;
  final NutritionSummary totalNutritionSummary;
  final Map<int, NutritionSummary> dailyNutritionBreakdown;
  final double estimatedTotalCostUsd;
  final double? budgetTargetUsd;
  final bool? isWithinBudget;
  final List<String> dietaryRestrictionsUsed;
  final String algorithmVersion;
  final int? userRating;
  final String? userFeedback;
  final DateTime createdAt;
  final DateTime? updatedAt;

  MealPlan({
    required this.id,
    required this.userId,
    required this.planDate,
    required this.durationDays,
    required this.meals,
    required this.totalNutritionSummary,
    required this.dailyNutritionBreakdown,
    required this.estimatedTotalCostUsd,
    this.budgetTargetUsd,
    this.isWithinBudget,
    required this.dietaryRestrictionsUsed,
    required this.algorithmVersion,
    this.userRating,
    this.userFeedback,
    required this.createdAt,
    this.updatedAt,
  });

  factory MealPlan.fromJson(Map<String, dynamic> json) {
    final totalCost = MealPlan._safeToDouble(json['estimated_total_cost_usd']) ?? 0.0;
    final mealsData = json['meals'] as List<dynamic>;
    
    return MealPlan(
      id: json['id'] as String,
      userId: json['user_id'] as String,
      planDate: DateTime.parse(json['plan_date'] as String),
      durationDays: json['duration_days'] as int,
      meals: mealsData
          .map((meal) => Meal.fromJson(meal as Map<String, dynamic>))
          .toList(),
      totalNutritionSummary: NutritionSummary.fromJson(
          json['total_nutrition_summary'] as Map<String, dynamic>),
      dailyNutritionBreakdown: _parseDailyNutrition(json['daily_nutrition_breakdown']),
      estimatedTotalCostUsd: totalCost,
      budgetTargetUsd: json['budget_target_usd'] != null
          ? MealPlan._safeToDouble(json['budget_target_usd'])
          : null,
      isWithinBudget: json['is_within_budget'] as bool?,
      dietaryRestrictionsUsed: (json['dietary_restrictions_used'] as List<dynamic>)
          .map((item) => item as String)
          .toList(),
      algorithmVersion: json['algorithm_version'] as String,
      userRating: json['user_rating'] as int?,
      userFeedback: json['user_feedback'] as String?,
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: json['updated_at'] != null
          ? DateTime.parse(json['updated_at'] as String)
          : null,
    );
  }

  static double? _safeToDouble(dynamic value) {
    if (value == null) return null;
    if (value is double) return value;
    if (value is int) return value.toDouble();
    if (value is String) return double.tryParse(value);
    return null;
  }

  static Map<int, NutritionSummary> _parseDailyNutrition(dynamic dailyNutrition) {
    if (dailyNutrition == null || dailyNutrition is! Map<String, dynamic>) {
      return {};
    }
    
    return dailyNutrition.map((key, value) {
      if (value is Map<String, dynamic>) {
        return MapEntry(
          int.parse(key),
          NutritionSummary.fromJson(value),
        );
      }
      return MapEntry(int.parse(key), NutritionSummary(calories: 0, protein: 0, carbs: 0, fat: 0));
    });
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'user_id': userId,
      'plan_date': planDate.toIso8601String(),
      'duration_days': durationDays,
      'meals': meals.map((meal) => meal.toJson()).toList(),
      'total_nutrition_summary': totalNutritionSummary.toJson(),
      'daily_nutrition_breakdown': dailyNutritionBreakdown
          .map((key, value) => MapEntry(key.toString(), value.toJson())),
      'estimated_total_cost_usd': estimatedTotalCostUsd,
      'budget_target_usd': budgetTargetUsd,
      'is_within_budget': isWithinBudget,
      'dietary_restrictions_used': dietaryRestrictionsUsed,
      'algorithm_version': algorithmVersion,
      'user_rating': userRating,
      'user_feedback': userFeedback,
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt?.toIso8601String(),
    };
  }

  DateTime get endDate => planDate.add(Duration(days: durationDays - 1));

  bool get isActive => DateTime.now().isBefore(endDate.add(const Duration(days: 1)));

  List<Meal> getMealsForDay(int day) {
    return meals.where((meal) => meal.day == day).toList();
  }

  double get averageDailyCost => estimatedTotalCostUsd / durationDays;

  bool get hasUserFeedback => userRating != null || userFeedback != null;
}

class Meal {
  final int day;
  final String mealType;
  final String recipeId;
  final String recipeName;
  final double score;
  final double? estimatedCostUsd;

  Meal({
    required this.day,
    required this.mealType,
    required this.recipeId,
    required this.recipeName,
    required this.score,
    this.estimatedCostUsd,
  });

  factory Meal.fromJson(Map<String, dynamic> json) {
    return Meal(
      day: json['day'] as int,
      mealType: json['meal_type'] as String,
      recipeId: json['recipe_id'] as String,
      recipeName: json['recipe_name'] as String,
      score: MealPlan._safeToDouble(json['score']) ?? 0.0,
      estimatedCostUsd: json['estimated_cost_usd'] != null 
          ? MealPlan._safeToDouble(json['estimated_cost_usd'])
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'day': day,
      'meal_type': mealType,
      'recipe_id': recipeId,
      'recipe_name': recipeName,
      'score': score,
      if (estimatedCostUsd != null) 'estimated_cost_usd': estimatedCostUsd,
    };
  }

  bool get isBreakfast => mealType == 'breakfast';
  bool get isLunch => mealType == 'lunch';
  bool get isDinner => mealType == 'dinner';
  bool get isSnack => mealType == 'snack';
}

class NutritionSummary {
  final double calories;
  final double protein;
  final double carbs;
  final double fat;
  final double? cost;

  NutritionSummary({
    required this.calories,
    required this.protein,
    required this.carbs,
    required this.fat,
    this.cost,
  });

  factory NutritionSummary.fromJson(Map<String, dynamic> json) {
    return NutritionSummary(
      calories: MealPlan._safeToDouble(json['calories']) ?? 0.0,
      protein: MealPlan._safeToDouble(json['protein']) ?? 0.0,
      carbs: MealPlan._safeToDouble(json['carbs']) ?? 0.0,
      fat: MealPlan._safeToDouble(json['fat']) ?? 0.0,
      cost: json['cost'] != null ? MealPlan._safeToDouble(json['cost']) : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'calories': calories,
      'protein': protein,
      'carbs': carbs,
      'fat': fat,
      if (cost != null) 'cost': cost,
    };
  }

  double get totalMacros => protein + carbs + fat;

  double get proteinPercentage => totalMacros > 0 ? (protein / totalMacros) * 100 : 0;
  double get carbsPercentage => totalMacros > 0 ? (carbs / totalMacros) * 100 : 0;
  double get fatPercentage => totalMacros > 0 ? (fat / totalMacros) * 100 : 0;
}

class MealPlanListItem {
  final String id;
  final DateTime planDate;
  final int durationDays;
  final int mealsCount;
  final double estimatedTotalCostUsd;
  final bool? isWithinBudget;
  final int? userRating;
  final DateTime createdAt;

  MealPlanListItem({
    required this.id,
    required this.planDate,
    required this.durationDays,
    required this.mealsCount,
    required this.estimatedTotalCostUsd,
    this.isWithinBudget,
    this.userRating,
    required this.createdAt,
  });

  factory MealPlanListItem.fromJson(Map<String, dynamic> json) {
    return MealPlanListItem(
      id: json['id'] as String,
      planDate: DateTime.parse(json['plan_date'] as String),
      durationDays: json['duration_days'] as int,
      mealsCount: json['meals_count'] as int,
      estimatedTotalCostUsd: MealPlan._safeToDouble(json['estimated_total_cost_usd']) ?? 0.0,
      isWithinBudget: json['is_within_budget'] as bool?,
      userRating: json['user_rating'] as int?,
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }

  DateTime get endDate => planDate.add(Duration(days: durationDays - 1));
  double get averageDailyCost => estimatedTotalCostUsd / durationDays;
} 

File: backend/src/core/models/meal_plan.py
"""
MealPlan Domain Model
Represents a generated meal plan for a user
"""

import uuid
from datetime import datetime, date
from typing import Optional, List, Dict, Any
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy import Column, String, DateTime, Date, Boolean, ForeignKey
from sqlalchemy.types import JSON

from data_access.database import db

class MealPlan(db.Model):
    """MealPlan model for storing generated meal plans"""
    
    __tablename__ = 'meal_plans'
    
    # Primary Fields
    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    user_id = Column(UUID(as_uuid=True), ForeignKey('users.id'), nullable=False)
    
    # Plan Configuration
    plan_date = Column(Date, nullable=False)  # Start date of the plan
    duration_days = Column(db.Integer, default=1)  # Number of days (1-7)
    
    # Meal Data - stored as JSON for flexibility
    meals = Column(JSON, nullable=False)  # [{"meal_type": "breakfast", "recipe_id": "xyz", "day": 1}]
    
    # Nutritional Summary
    total_nutrition_summary = Column(JSON, nullable=True)  # {"calories": 2000, "protein": 100, "carbs": 250, "fat": 65}
    daily_nutrition_breakdown = Column(JSON, nullable=True)  # Per-day nutrition breakdown
    
    # Generation Metadata
    generated_by_ai = Column(Boolean, default=True)
    generation_parameters = Column(JSON, nullable=True)  # Parameters used for generation
    algorithm_version = Column(String(50), nullable=True)  # Version of algorithm used
    
    # Budget and Preferences
    estimated_total_cost_usd = Column(db.Integer, nullable=True)  # Total cost in cents
    budget_target_usd = Column(db.Integer, nullable=True)  # Target budget in cents
    dietary_restrictions_used = Column(JSON, nullable=True)  # Restrictions applied during generation
    
    # Status and Feedback
    is_active = Column(Boolean, default=True)
    user_rating = Column(db.Integer, nullable=True)  # 1-5 rating from user
    user_feedback = Column(String(1000), nullable=True)  # Optional feedback text
    
    # Timestamps
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
    
    def __init__(self, user_id: str, plan_date: date, meals: List[Dict[str, Any]],
                 duration_days: int = 1, total_nutrition_summary: Optional[Dict[str, Any]] = None,
                 daily_nutrition_breakdown: Optional[Dict[str, Any]] = None,
                 generated_by_ai: bool = True, generation_parameters: Optional[Dict[str, Any]] = None,
                 algorithm_version: Optional[str] = None, estimated_total_cost_usd: Optional[int] = None,
                 budget_target_usd: Optional[int] = None, dietary_restrictions_used: Optional[List[str]] = None):
        """Initialize a new meal plan"""
        self.user_id = user_id
        self.plan_date = plan_date
        self.duration_days = duration_days
        self.meals = meals
        self.total_nutrition_summary = total_nutrition_summary or {}
        self.daily_nutrition_breakdown = daily_nutrition_breakdown or {}
        self.generated_by_ai = generated_by_ai
        self.generation_parameters = generation_parameters or {}
        self.algorithm_version = algorithm_version
        self.estimated_total_cost_usd = estimated_total_cost_usd
        self.budget_target_usd = budget_target_usd
        self.dietary_restrictions_used = dietary_restrictions_used or []
    
    @property
    def end_date(self) -> date:
        """Calculate the end date of the meal plan"""
        from datetime import timedelta
        return self.plan_date + timedelta(days=self.duration_days - 1)
    
    @property
    def cost_per_day_usd(self) -> Optional[float]:
        """Calculate average cost per day in USD"""
        if self.estimated_total_cost_usd and self.duration_days > 0:
            return (self.estimated_total_cost_usd / 100.0) / self.duration_days
        return None
    
    @property
    def calories_per_day(self) -> Optional[float]:
        """Calculate average calories per day"""
        if self.total_nutrition_summary and 'calories' in self.total_nutrition_summary:
            return self.total_nutrition_summary['calories'] / self.duration_days
        return None
    
    @property
    def is_within_budget(self) -> Optional[bool]:
        """Check if meal plan is within budget (±10% tolerance)"""
        if not self.estimated_total_cost_usd or not self.budget_target_usd:
            return None
        
        tolerance = 0.10
        budget_min = self.budget_target_usd * (1 - tolerance)
        budget_max = self.budget_target_usd * (1 + tolerance)
        
        return budget_min <= self.estimated_total_cost_usd <= budget_max
    
    def get_meals_by_day(self, day: int) -> List[Dict[str, Any]]:
        """Get all meals for a specific day"""
        return [meal for meal in self.meals if meal.get('day') == day]
    
    def get_meals_by_type(self, meal_type: str) -> List[Dict[str, Any]]:
        """Get all meals of a specific type across all days"""
        return [meal for meal in self.meals if meal.get('meal_type') == meal_type]
    
    def get_unique_recipes(self) -> List[str]:
        """Get list of unique recipe IDs used in this meal plan"""
        return list(set(meal.get('recipe_id') for meal in self.meals if meal.get('recipe_id')))
    
    def calculate_variety_score(self) -> float:
        """Calculate variety score based on unique recipes vs total meals"""
        if not self.meals:
            return 0.0
        
        unique_recipes = len(self.get_unique_recipes())
        total_meals = len(self.meals)
        
        return unique_recipes / total_meals if total_meals > 0 else 0.0
    
    def add_user_feedback(self, rating: int, feedback: Optional[str] = None) -> None:
        """Add user rating and feedback"""
        if not 1 <= rating <= 5:
            raise ValueError("Rating must be between 1 and 5")
        
        self.user_rating = rating
        self.user_feedback = feedback
        self.updated_at = datetime.utcnow()
    
    def validate_meal_plan_structure(self) -> bool:
        """Validate that the meal plan has proper structure"""
        if not self.meals:
            return False
        
        required_meal_fields = ['meal_type', 'recipe_id', 'day']
        valid_meal_types = ['breakfast', 'lunch', 'dinner', 'snack']
        
        for meal in self.meals:
            # Check required fields
            if not all(field in meal for field in required_meal_fields):
                return False
            
            # Check valid meal type
            if meal['meal_type'] not in valid_meal_types:
                return False
            
            # Check valid day range
            if not 1 <= meal['day'] <= self.duration_days:
                return False
        
        return True
    
    def to_dict(self, include_detailed_nutrition: bool = False) -> Dict[str, Any]:
        """Convert meal plan to dictionary for API responses"""
        result = {
            'id': str(self.id),
            'user_id': str(self.user_id),
            'plan_date': self.plan_date.isoformat(),
            'end_date': self.end_date.isoformat(),
            'duration_days': self.duration_days,
            'meals': self.meals,
            'total_nutrition_summary': self.total_nutrition_summary,
            'generated_by_ai': self.generated_by_ai,
            'algorithm_version': self.algorithm_version,
            'estimated_total_cost_usd': self.estimated_total_cost_usd,
            'cost_per_day_usd': self.cost_per_day_usd,
            'budget_target_usd': self.budget_target_usd,
            'is_within_budget': self.is_within_budget,
            'calories_per_day': self.calories_per_day,
            'dietary_restrictions_used': self.dietary_restrictions_used,
            'variety_score': self.calculate_variety_score(),
            'is_active': self.is_active,
            'user_rating': self.user_rating,
            'user_feedback': self.user_feedback,
            'created_at': self.created_at.isoformat() if self.created_at else None,
            'updated_at': self.updated_at.isoformat() if self.updated_at else None,
        }
        
        if include_detailed_nutrition:
            result['daily_nutrition_breakdown'] = self.daily_nutrition_breakdown
            result['generation_parameters'] = self.generation_parameters
        
        return result
    
    def __repr__(self) -> str:
        return f'<MealPlan {self.id} for user {self.user_id} ({self.plan_date} - {self.duration_days} days)>' 